btfgen.c

```diff
@@ -274,56 +342,107 @@ static struct btf_reloc_type *btf_reloc_put_type(struct btf *btf,
+						     struct btf_reloc_info *info,
+						     struct btf_type *btf_type,
+						     unsigned int id) {
+	int err, i, n;
+	unsigned int child_id;
+	struct btf_type *btf_tmp;
+	struct btf_array *array;
+	struct btf_member *m;
+	struct btf_reloc_type *reloc_tmp;
+	struct btf_reloc_type *reloc_type;
 
-	if (!hashmap__find(info->types, uint_as_hash_key(id), (void **)&type))
-		return ERR_PTR(-ENOENT);
+	if (!hashmap__find(info->types, uint_as_hash_key(id), (void**) &reloc_type)) {
+		reloc_type = calloc(1, sizeof(*reloc_type));
+		if (!reloc_type)
+			return ERR_PTR(-ENOMEM);
 
-	return type;
-}
+		reloc_type->type = btf_type;
+		reloc_type->id = id;
 
-static int bpf_reloc_type_add_member(struct btf_reloc_info *info,
-				     struct btf_reloc_type *reloc_type,
-				     struct btf_member *btf_member, int idx) {
-	int err;
-	struct btf_reloc_member *reloc_member;
+		err = hashmap__add(info->types, uint_as_hash_key(reloc_type->id), reloc_type);
+		if (err)
+			return ERR_PTR(err);
+	}
 
-	/* create new members hashmap for this relocation type if needed */
-	if (reloc_type->members == NULL) {
-		struct hashmap *tmp = hashmap__new(bpf_reloc_info_hash_fn,
-						   bpf_reloc_info_equal_fn,
-						   NULL);
-		if (IS_ERR(tmp))
-			return PTR_ERR(tmp);
+	switch (btf_kind(reloc_type->type)) {
+	case BTF_KIND_INT:
+	case BTF_KIND_FLOAT:
+		/* not a complex type, already solved */
+		break;
+	case BTF_KIND_STRUCT:
+	case BTF_KIND_UNION:
+		n = btf_vlen(btf_type);
+		m = btf_members(btf_type);
+		for (i = 0; i < n; i++, m++) {
+			btf_type = (struct btf_type *) btf__type_by_id(btf, m->type);
+			if (!btf_type)
+				return ERR_PTR(-EINVAL);
+
+			/* add all member types */
+			reloc_tmp = btf_reloc_put_type_all(btf, info, btf_type, m->type);
+			if (IS_ERR(reloc_tmp))
+				return reloc_tmp;
+
+			/* add all members */
+                        err = bpf_reloc_type_add_member(info, reloc_type, m, i);
+                        if (err)
+				return ERR_PTR(err);
+		}
+		break;
+	case BTF_KIND_PTR:
+	case BTF_KIND_CONST:
+	case BTF_KIND_VOLATILE:
+	case BTF_KIND_TYPEDEF:
+		/* modifier types */
+		child_id = btf_type->type;
+		btf_type = (struct btf_type *) btf__type_by_id(btf, child_id);
+		if (!btf_type)
+			return ERR_PTR(-EINVAL);
 
-		reloc_type->members = tmp;
-	}
-	/* add given btf_member as a member of the parent relocation_type's type */
-	reloc_member = calloc(1, sizeof(*reloc_member));
-	if (!reloc_member)
-		return -ENOMEM;
-	reloc_member->member = btf_member;
-	reloc_member->idx = idx;
-	/* add given btf_member as member to given relocation type */
-	err = hashmap__add(reloc_type->members, uint_as_hash_key(reloc_member->idx), reloc_member);
-	if (err) {
-		free(reloc_member);
-		if (err != -EEXIST)
-			return err;
+		reloc_tmp = btf_reloc_put_type_all(btf, info, btf_type, child_id);
+		if (IS_ERR(reloc_tmp))
+			return reloc_tmp;
+		break;
+	case BTF_KIND_ARRAY:
+		child_id = btf_type->type;
```

Comment:
I'm not sure btf_type->type is relevant for an array: https://github.com/libbpf/libbpf/blob/3ef05a585efd47abf4dc92265430d0248d7f388a/include/uapi/linux/btf.h#L45. I'd say that `array->type` and `array->index_type` should be used instead.

Reply:
Nope, its not. My array handling was poorly designed and untested. Thanks for the review and corrections.